pkg/cmd/cli/cmd/rsync.go

```diff
@@ -0,0 +1,372 @@
+	flags := "-a"
+	if !o.Quiet {
+		flags += "v"
+	}
+	destination := fmt.Sprintf("%s:%s", o.PodName, o.Destination)
+	cmd := []string{defaultRsyncExecutable, flags}
+	if o.Delete {
+		cmd = append(cmd, "--delete")
+	}
+	cmd = append(cmd, "-e", o.RshCommand, o.Source, destination)
+	glog.V(4).Infof("Local command: %s", strings.Join(cmd, " "))
+	return o.LocalExecutor.Execute(cmd, nil, o.Out, o.ErrOut)
+}
+
+func (o *RsyncOptions) copyWithTar() error {
+	glog.V(3).Infof("Copying files with tar")
+	if o.Delete {
+		deleteCmd := []string{"sh", "-c", fmt.Sprintf("rm -rf %s", filepath.Join(o.Destination, "*"))}
+		err := executeWithLogging(o.RemoteExecutor, deleteCmd)
+		if err != nil {
+			return fmt.Errorf("unable to delete files in destination: %v", err)
+		}
+	}
+	tmp, err := ioutil.TempFile("", "rsync")
+	if err != nil {
+		return err
+	}
+	err = tarLocal(o.Tar, o.Source, tmp)
+	if err != nil {
+		return err
+	}
+	err = tmp.Close()
+	if err != nil {
+		return err
+	}
+	tmp, err = os.Open(tmp.Name())
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tmp.Name())
+
+	flags := "x"
+	if !o.Quiet {
+		flags += "v"
+	}
+	remoteCmd := []string{defaultTarExecutable, flags, "-C", o.Destination}
+	return o.RemoteExecutor.Execute(remoteCmd, tmp, o.Out, o.ErrOut)
+}
+
+func tarLocal(tar tar.Tar, sourceDir string, w io.Writer) error {
+	glog.V(4).Infof("Tarring %s locally", sourceDir)
+	// includeParent mimics rsync's behavior. When the source path ends in a path
+	// separator, then only the contents of the directory are copied. Otherwise,
+	// the directory itself is copied.
+	includeParent := true
+	if strings.HasSuffix(sourceDir, string(filepath.Separator)) {
+		includeParent = false
+		sourceDir = sourceDir[:len(sourceDir)-1]
+	}
+	return tar.CreateTarStream(sourceDir, includeParent, w)
+}
+
+// RunRsync copies files from source to destination
+func (o *RsyncOptions) RunRsync() error {
+
+	if !o.UseTar && o.checkLocalRsync() == nil {
+		// If rsync is present in the pod's container, then use it
+		if o.checkRemoteRsync() == nil {
+			return o.copyWithRsync()
+		}
+	}
+
+	// If tar is present in the pod's container, then use it
+	if o.checkRemoteTar() == nil {
+		if !o.UseTar {
+			glog.Warningf("Could not find rsync. It must be on the path of the local machine and of the remote container, using tar.")
+		}
+		return o.copyWithTar()
+	}
+
+	return fmt.Errorf("the container in pod %s/%s does not have rsync or tar", o.Namespace, o.PodName)
+}
+
+func executeWithLogging(e executor, cmd []string) error {
+	w := &bytes.Buffer{}
+	err := e.Execute(cmd, nil, w, w)
+	glog.V(4).Infof("%s", w.String())
+	glog.V(4).Infof("error: %v", err)
+	return err
+}
+
+func (o *RsyncOptions) checkLocalRsync() error {
+	_, err := exec.LookPath("rsync")
+	if err != nil {
+		glog.V(3).Infof("local rsync not found: %v", err)
+	}
+	return err
+}
+
+func (o *RsyncOptions) checkRemoteRsync() error {
+	err := executeWithLogging(o.RemoteExecutor, testRsyncCommand)
```

Comment:
It takes 100ms on average. We could also check the return code (14 is IPC error) and error output of the local rsync call. If we can determine that it's because rsync is not present in the container, then fallback to tar.